search/processor.cpp

```diff
@@ -376,59 +335,51 @@ int Processor::GetCategoryLocales(int8_t(&arr)[3]) const
   return count;
 }
 
-void Processor::ForEachCategoryType(StringSliceBase const & slice,
-                                     function<void(size_t, uint32_t)> const & fn) const
+template <typename ToDo>
+void Processor::ForEachCategoryType(StringSliceBase const & slice, ToDo && todo) const
 {
   int8_t arrLocales[3];
   int const localesCount = GetCategoryLocales(arrLocales);
 
-  for (size_t i = 0; i < slice.Size(); ++i)
-  {
-    auto token = RemoveHashtag(slice.Get(i));
-    for (int j = 0; j < localesCount; ++j)
-      m_categories.ForEachTypeByName(arrLocales[j], token, bind<void>(fn, i, _1));
-    ProcessEmojiIfNeeded(token, i, fn);
-  }
-}
-
-// template <class ToDo>
-void Processor::ProcessEmojiIfNeeded(strings::UniString const & token, size_t index,
-                                     function<void(size_t, uint32_t)> const & fn) const
-{
-  // Special process of 2 codepoints emoji (e.g. black guy on a bike).
-  // Only emoji synonyms can have one codepoint.
-  if (token.size() > 1)
-  {
-    static int8_t const enLocaleCode = CategoriesHolder::MapLocaleToInteger("en");
-
-    m_categories.ForEachTypeByName(enLocaleCode, strings::UniString(1, token[0]),
-                                   bind<void>(fn, index, _1));
-  }
+  ::search::ForEachCategoryType(slice, arrLocales, localesCount, m_categories, forward<ToDo>(todo));
 }
 
 void Processor::Search(Results & results, size_t limit)
 {
-  if (m_tokens.empty())
-    SuggestStrings(results);
+  Geocoder::Params geocoderParams;
+  InitParams(geocoderParams);
+  geocoderParams.m_mode = m_mode;
+  geocoderParams.m_pivot = GetPivotRect();
+  geocoderParams.m_accuratePivotCenter = GetPivotPoint();
+  m_geocoder.SetParams(geocoderParams);
+
+  Ranker::Params rankerParams;
+  rankerParams.m_currentLocaleCode = m_currentLocaleCode;
+  if (m_mode == Mode::Viewport)
```

Comment:
On the other hand, we probably need to get rid of public SearchViewportPoints() and Search() functions. Currently SearchEngine dispatches calls to Processor. Probably we should do this dispatching inside Processor.